packages/neos-ui/src/Sagas/Changes/index.js

```diff
@@ -1,34 +1,72 @@
 import backend from '@neos-project/neos-ui-backend-connector';
 
 const {publishableNodesInDocumentSelector, baseWorkspaceSelector} = selectors.CR.Workspaces;
 
-export function * watchPersist() {
+const saveDelay = 2000;
+
+function * persistChanges(changes) {
     const {change} = backend.get().endpoints;
 
-    yield takeEvery(actionTypes.Changes.PERSIST, function * persistChanges(action) {
-        const changes = action.payload.changes;
+    yield put(actions.UI.Remote.startSaving());
+
+    try {
+        const feedback = yield call(change, changes);
+        yield put(actions.UI.Remote.finishSaving());
+        yield put(actions.ServerFeedback.handleServerFeedback(feedback));
+
+        const state = yield select();
+        const isAutoPublishingEnabled = $get('user.settings.isAutoPublishingEnabled', state);
+
+        if (isAutoPublishingEnabled) {
+            const baseWorkspace = baseWorkspaceSelector(state);
+            const publishableNodesInDocument = publishableNodesInDocumentSelector(state);
+            yield put(actions.CR.Workspaces.publish(publishableNodesInDocument.map($get('contextPath')), baseWorkspace));
+        }
+    } catch (error) {
+        console.error('Failed to persist changes', error);
+    } finally {
+        yield put(actions.UI.Remote.finishSaving());
+    }
+}
+
+export function * watchPersist() {
+    let changes = [];
+    let lastPush = Date.now();
+    let actionsTaken = 0;
+
+    while (true) {
+        const now = Date.now();
+        const timeSinceLastPush = now - lastPush;
 
-        yield put(actions.UI.Remote.startSaving());
+        const {action} = yield race({
+            action: take(actionTypes.Changes.PERSIST),
+            time: call(delay, saveDelay)
+        });
 
-        try {
-            const feedback = yield call(change, changes);
-            yield put(actions.UI.Remote.finishSaving());
-            yield put(actions.ServerFeedback.handleServerFeedback(feedback));
+        if (action) {
+            changes.push(...action.payload.changes);
+            actionsTaken++;
+        }
 
-            const state = yield select();
-            const isAutoPublishingEnabled = $get('user.settings.isAutoPublishingEnabled', state);
+        const state = yield select();
 
-            if (isAutoPublishingEnabled) {
-                const baseWorkspace = baseWorkspaceSelector(state);
-                const publishableNodesInDocument = publishableNodesInDocumentSelector(state);
-                yield put(actions.CR.Workspaces.publish(publishableNodesInDocument.map($get('contextPath')), baseWorkspace));
-            }
-        } catch (error) {
-            console.error('Failed to persist changes', error);
+        /*
+         *  We actually persist IF:
+         *  - we have changes AND
+         *  - there is not another request running
+         *  - either the "saveDelay" time has passed OR we have taken 3 actions (to avoid huge change queues locally)
```

Comment:
Just wondering, how did you come up with this number of 3? Did you test having longer queue? Modern JS engines are really powerful and keeping even a few MBs in memory would be nothing.

Reply:
On the other hand I don't understand the need for `2s` delay here? What does it give us? The inputs that require delay should have already been debounced.

Reply:
yes for inline that is true, for inspector not so much it seemed (although I didn't check the code yet). It just seemed sensible to "collect" changes happening within a certain timeframe.